homeassistant/components/lastfm/coordinator.py

```diff
@@ -0,0 +1,90 @@
+"""DataUpdateCoordinator for the LastFM integration."""
+from __future__ import annotations
+
+from dataclasses import dataclass
+from datetime import timedelta
+
+from pylast import LastFMNetwork, PyLastError, Track
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_API_KEY
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import (
+    CONF_USERS,
+    DOMAIN,
+    LOGGER,
+)
+
+
+def format_track(track: Track | None) -> str | None:
+    """Format the track."""
+    if track is None:
+        return None
+    return f"{track.artist} - {track.title}"
+
+
+@dataclass
+class LastFMUserData:
+    """Data holder for LastFM data."""
+
+    play_count: int
+    image: str
+    now_playing: str | None
+    top_track: str | None
+    last_track: str | None
+
+
+class LastFMDataUpdateCoordinator(
+    DataUpdateCoordinator[dict[str, LastFMUserData | None]]
+):
+    """A LastFM Data Update Coordinator."""
+
+    config_entry: ConfigEntry
+    _client: LastFMNetwork
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the LastFM data coordinator."""
+        super().__init__(
+            hass,
+            LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=30),
+        )
+        self._client = LastFMNetwork(api_key=self.config_entry.options[CONF_API_KEY])
+
+    async def _async_update_data(self) -> dict[str, LastFMUserData | None]:
+        res = {}
+        for username in self.config_entry.options[CONF_USERS]:
+            res[username] = await self.hass.async_add_executor_job(
+                self._get_user_data, username
+            )
+        if len(res.values()) == list(res.values()).count(None):
```

Comment:
```suggestion
        if not res:
```

Reply:
Currently the dict would look `{"username": None}` if a user is not found, so this wouldn't work right?

Reply:
Not exactly but it would save to the dict so change it so it only adds to the dict if `_get_user_data` actually returns data